promql/engine.go

```diff
@@ -777,6 +780,7 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 			seriess[h] = ss
 
 		}
+		ev.currentSamples = tempNumSamples
```

Comment:
This can return early up a bit, so this needs reworking.

The input vectors to the function also need accounting for when they're no longer in use.

Reply:
> The input vectors to the function also need accounting for when they're no longer in use.

Not sure what you mean here. We should reset the number of samples immediately after the call to `f` rather than later on?

Reply:
With recursive calls to eval, once the inner evals are done we don't need to account for their memory.